pkg/cloud/services/ec2/instances.go

```diff
@@ -226,30 +226,63 @@ func (s *Service) CreateInstance(scope *scope.MachineScope, userData []byte) (*i
 // - subnet based on filters in machine configuration
 // - subnet based on the availability zone specified,
 // - default to the first private subnet available.
+// Filters above subnets for public if .publicIP is specified in machine spec.
 func (s *Service) findSubnet(scope *scope.MachineScope) (string, error) {
 	// Check Machine.Spec.FailureDomain first as it's used by KubeadmControlPlane to spread machines across failure domains.
 	failureDomain := scope.Machine.Spec.FailureDomain
 	if failureDomain == nil {
 		failureDomain = scope.AWSMachine.Spec.FailureDomain
 	}
 
+	wantPublicIP := aws.BoolValue(scope.AWSMachine.Spec.PublicIP)
+	var subnets infrav1.Subnets
+
 	switch {
 	case scope.AWSMachine.Spec.Subnet != nil && scope.AWSMachine.Spec.Subnet.ID != nil:
-		return *scope.AWSMachine.Spec.Subnet.ID, nil
-	case scope.AWSMachine.Spec.Subnet != nil && scope.AWSMachine.Spec.Subnet.Filters != nil:
-		subnetInput := &ec2.DescribeSubnetsInput{
+		subnet := scope.AWSMachine.Spec.Subnet
+
+		// Describe subnet to check for public IP support.
+		out, err := s.EC2Client.DescribeSubnets(&ec2.DescribeSubnetsInput{
+			SubnetIds: []*string{subnet.ID},
 			Filters: []*ec2.Filter{
-				filter.EC2.SubnetStates(ec2.SubnetStatePending, ec2.SubnetStateAvailable),
-				filter.EC2.VPC(s.scope.VPC().ID),
+				filter.EC2.PublicSubnets(),
```

Comment:
Does this not result in filtering for public subnets regardless of whether or not the user wanted or not?

I'm slightly cautious of adding extraneous EC2 calls, so can we do the following:
* Break this logic out to a separate function so it's easier to read
* Conditionally do the check based on whether or not the user specified wanting public IP support or not
* add test cases for the new function

Reply:
@dilyevsky have you had a chance to see @randomvariable's question above?

Reply:
> Does this not result in filtering for public subnets regardless of whether or not the user wanted or not?

@randomvariable good catch, thanks - I need extra condition for `len(out.Subnets) == 0` check.

> I'm slightly cautious of adding extraneous EC2 calls, so can we do the following:

I'll look into how I can possibly cache the result from previous calls...

Working on tests now